#### Author: Mustafa Radheyyan
#### Date: 05/01/2023
#### Assignment: Cognixia JUMPlus Python Project 5
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

## Import required libraries

In [1]:
import pandas as pd

Path of data source

In [2]:
ECOMMERCE_PURCHASES_CSV_PATH = 'Ecommerce_purchases.csv'

Context Manager helper functions for displaying DataFrames differently than the default configuration settings

In [3]:
def force_show_all_columns(df):
    with pd.option_context('display.max_columns', None, 'display.width', None):
        display(df)

In [4]:
def force_show_all_rows(df):
    with pd.option_context('display.max_rows', None, 'display.width', None):
        display(df)

In [5]:
def use_inf_as_na(df, df_command):
    with pd.option_context('mode.use_inf_as_na', True):
        return df_command(df)

## Read the csv file

There are no specific index columns in this dataset.

In [6]:
ecommerce_purchases_df = pd.read_csv(ECOMMERCE_PURCHASES_CSV_PATH)

## Data Cleaning

##### Convert DF Column Name List

Made DataFrame's columns snake case in order to be more computing friendly<br></br>
Made a list of the dataframe's columns to enable easier searching of column names

In [7]:
ecommerce_purchases_df.columns = ecommerce_purchases_df.columns.str.replace(' ', '_').str.lower()
ecommerce_purchases_columns = list(ecommerce_purchases_df.columns)
ecommerce_purchases_columns

['address',
 'lot',
 'am_or_pm',
 'browser_info',
 'company',
 'credit_card',
 'cc_exp_date',
 'cc_security_code',
 'cc_provider',
 'email',
 'job',
 'ip_address',
 'language',
 'purchase_price']

#### Data Cleaning

In [8]:
ecommerce_purchases_df.loc[0, 'address']

'16629 Pace Camp Apt. 448\r\nAlexisborough, NE 77130-7478'

In [9]:
ecommerce_purchases_df['address'] = ecommerce_purchases_df['address'].str.replace('\n', '')

In [10]:
ecommerce_purchases_df['am_or_pm'].unique()

array(['PM', 'AM'], dtype=object)

In [11]:
def am_or_pm_conversion(cell_value):
    if cell_value == "AM":
        return True
    elif cell_value == "PM":
        return False


In [12]:
ecommerce_purchases_df['am_or_pm'] = ecommerce_purchases_df['am_or_pm'].apply(am_or_pm_conversion)

In [13]:
len(ecommerce_purchases_df['lot'].unique())

9822

In [14]:
ecommerce_purchases_df = ecommerce_purchases_df.rename(columns={'am_or_pm': 'time_am'})

In [15]:
ecommerce_purchases_df = ecommerce_purchases_df.convert_dtypes()

In [16]:
ecommerce_purchases_columns = list(ecommerce_purchases_df.columns)

## 1. Display Top 10 Rows of The Dataset

In [17]:
ecommerce_purchases_df.head(10)

address    lot  time_am   
0  16629 Pace Camp Apt. 448
Alexisborough, NE 771...  46 in    False  \
1  9374 Jasmine Spurs Suite 508
South John, TN 84...  28 rn    False   
2                    Unit 0065 Box 5052
DPO AP 27450  94 vE    False   
3               7780 Julia Fords
New Stacy, WA 45798  36 vm    False   
4  23012 Munoz Drive Suite 337
New Cynthia, TX 57826  20 IE     True   
5  7502 Powell Mission Apt. 768
Travisland, VA 30...  21 XT    False   
6      93971 Conway Causeway
Andersonburgh, AZ 75107  96 Xt     True   
7  260 Rachel Plains Suite 366
Castroberg, WV 248...  96 pG    False   
8             2129 Dylan Burg
New Michelle, ME 28650  45 JN    False   
9     3795 Dawson Extensions
Lake Tinafort, ID 88739  15 Ug     True   

                                        browser_info   
0  Opera/9.56.(X11; Linux x86_64; sl-SI) Presto/2...  \
1  Opera/8.93.(Windows 98; Win 9x 4.90; en-US) Pr...   
2  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
3  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_0 ...   
4  Opera/9.58.(X11; Linux x86_64; it-IT) Presto/2...   
5  Mozilla/5.0 (Macintosh; U; PPC Mac OS X 10_8_5...   
6  Mozilla/5.0 (compatible; MSIE 7.0; Windows NT ...   
7  Mozilla/5.0 (X11; Linux i686) AppleWebKit/5350...   
8  Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_7...   
9  Mozilla/5.0 (X11; Linux i686; rv:1.9.7.20) Gec...   

                           company       credit_card cc_exp_date   
0                  Martinez-Herman  6011929061123406       02/20  \
1  Fletcher, Richards and Whitaker  3337758169645356       11/18   
2       Simpson, Williams and Pham      675957666125       08/19   
3  Williams, Marshall and Buchanan  6011578504430710       02/24   
4        Brown, Watson and Andrews  6011456623207998       10/25   
5                   Silva-Anderson    30246185196287       07/25   
6                  Gibson and Sons  6011398782655569       07/24   
7                 Marshall-Collins      561252141909       06/25   
8                Galloway and Sons   180041795790001       04/24   
9     Rivera, Buchanan and Ramirez     4396283918371       01/17   

   cc_security_code                  cc_provider   
0               900                 JCB 16 digit  \
1               561                   Mastercard   
2               699                 JCB 16 digit   
3               384                     Discover   
4               678  Diners Club / Carte Blanche   
5              7169                     Discover   
6               714                VISA 16 digit   
7               256                VISA 13 digit   
8               899                 JCB 16 digit   
9               931             American Express   

                            email                                     job   
0               pdunlap@yahoo.com  Scientist, product/process development  \
1              anthony41@reed.com                       Drilling engineer   
2  amymiller@morales-harrison.com                Customer service manager   
3     brent16@olson-robinson.info                       Drilling engineer   
4     christopherwright@gmail.com                             Fine artist   
5               ynguyen@gmail.com                       Fish farm manager   
6              olivia04@yahoo.com                                  Dancer   
7            phillip48@parks.info                         Event organiser   
8            kdavis@rasmussen.com                       Financial manager   
9        qcoleman@hunt-huerta.com                      Forensic scientist   

        ip_address language  purchase_price  
0  149.146.147.205       el           98.14  
1     15.160.41.51       fr           70.73  
2   132.207.160.22       de            0.95  
3     30.250.74.19       es           78.04  
4     24.140.33.94       es           77.82  
5    55.96.152.147       ru           25.15  
6   127.252.144.18       de           88.56  
7   224.247.97.150       pt           44.25  
8  146.234.201.229       ru           59.54  
9    

## 2. Check Last 10 Rows of The Dataset

In [18]:
ecommerce_purchases_df.tail(10)

address    lot  time_am   
9990   75731 Molly Springs
West Danielle, VT 96934-5102  93 ty    False  \
9991               PSC 8165, Box 8498
APO AP 60327-0346  50 dA     True   
9992   885 Allen Mountains Apt. 230
Wallhaven, LA 16995  40 vH    False   
9993  7555 Larson Locks Suite 229
Ellisburgh, MA 349...  72 jg    False   
9994        6276 Rojas Hollow
Lake Louis, WY 56410-7837  93 Ex    False   
9995       966 Castaneda Locks
West Juliafurt, CO 96415  92 XI    False   
9996  832 Curtis Dam Suite 785
North Edwardburgh, TX...  41 JY     True   
9997               Unit 4434 Box 6343
DPO AE 28026-0283  74 Zh     True   
9998                0096 English Rest
Roystad, IA 12457  74 cL    False   
9999      40674 Barrett Stravenue
Grimesville, WI 79682  64 Hr     True   

                                           browser_info   
9990  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_4;...  \
9991  Mozilla/5.0 (compatible; MSIE 8.0; Windows NT ...   
9992  Mozilla/5.0 (Macintosh; PPC Mac OS X 10_6_5) A...   
9993  Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_8...   
9994  Opera/9.68.(X11; Linux x86_64; sl-SI) Presto/2...   
9995  Mozilla/5.0 (Windows NT 5.1) AppleWebKit/5352 ...   
9996  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
9997  Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_7...   
9998  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_8;...   
9999  Mozilla/5.0 (X11; Linux i686; rv:1.9.5.20) Gec...   

                         company       credit_card cc_exp_date   
9990  Pace, Vazquez and Richards   869968197049750       04/24  \
9991                  Snyder Inc  4221582137197481       02/24   
9992                   Wells Ltd  4664825258997302       10/20   
9993              Colon and Sons    30025560104631       10/25   
9994                Ritter-Smith  3112186784121077       01/25   
9995               Randall-Sloan   342945015358701       03/22   
9996    Hale, Collins and Wilson   210033169205009       07/25   
9997                Anderson Ltd  6011539787356311       05/21   
9998                    Cook Inc   180003348082930       11/17   
9999                  Greene Inc  4139972901927273       02/19   

      cc_security_code       cc_provider                        email   
9990               877      JCB 15 digit  andersonmichael@sherman.biz  \
9991               969           Voyager           kking@wise-liu.com   
9992               431          Discover            bberry@wright.net   
9993               629           Maestro    chelseawilliams@lopez.biz   
9994              1823           Maestro           iroberts@gmail.com   
9995               838      JCB 15 digit       iscott@wade-garner.com   
9996               207      JCB 16 digit           mary85@hotmail.com   
9997                 1     VISA 16 digit            tyler16@gmail.com   
9998               987  American Express      elizabethmoore@reid.net   
9999               302      JCB 15 digit        rachelford@vaughn.com   

                               job       ip_address language  purchase_price  
9990           Early years teacher     54.170.3.185       ru           18.35  
9991         IT sales professional    254.25.31.156       el           25.93  
9992                  Set designer    174.173.51.32       de           67.96  
9993  Designer, exhibition/display    177.46.82.128       el           65.61  
9994     Education officer, museum    242.44.112.18       zh           31.85  
9995                    Printmaker    29.73.197.114       it           82.21  
9996               Energy engineer   121.133.168.51       pt           25.63  
9997            Veterinary surgeon    156.210.0.254       el           83.98  
9998      Local government officer     55.78.26.143       es           38.84  
9999        Embryologist, clinical  176.119.198.199       el           67.59

## 3. Check Datatype of Each Column

In [19]:
display(ecommerce_purchases_df.dtypes)

address             string[python]
lot                 string[python]
time_am                    boolean
browser_info        string[python]
company             string[python]
credit_card                  Int64
cc_exp_date         string[python]
cc_security_code             Int64
cc_provider         string[python]
email               string[python]
job                 string[python]
ip_address          string[python]
language            string[python]
purchase_price             Float64
dtype: object

## 4. Check null values in the dataset

### a. When we write isnull() method it creates boolean table

In [20]:
force_show_all_columns(ecommerce_purchases_df.isna().any().to_frame(name='Has NaN or N/A').transpose())

address    lot  time_am  browser_info  company  credit_card   
Has NaN or N/A    False  False    False         False    False        False  \

                cc_exp_date  cc_security_code  cc_provider  email    job   
Has NaN or N/A        False             False        False  False  False  \

                ip_address  language  purchase_price  
Has NaN or N/A       False     False           False

### b. Now you have to count that is there any missing value in the dataset or not ?

In [21]:
null_ecommerce_purchases_df = (ecommerce_purchases_df.isna() | ecommerce_purchases_df.eq('')).astype('bool')

null_ecommerce_purchases_count_df = (
    null_ecommerce_purchases_df
    .sum(axis=0)
    .to_frame(name='# of null, empty, N/A')
    .transpose()
)

force_show_all_columns(null_ecommerce_purchases_count_df)

address  lot  time_am  browser_info  company   
# of null, empty, N/A        0    0        0             0        0  \

                       credit_card  cc_exp_date  cc_security_code   
# of null, empty, N/A            0            0                 0  \

                       cc_provider  email  job  ip_address  language   
# of null, empty, N/A            0      0    0           0         0  \

                       purchase_price  
# of null, empty, N/A               0

In [22]:
num_of_missing_values_in_dataset = list(null_ecommerce_purchases_count_df.sum(axis=1).values)[0]
num_of_missing_values_in_dataset

0

In [23]:
from IPython.display import HTML

display(HTML(f"""<h1>There {'are no missing values' if num_of_missing_values_in_dataset <= 0
    else 'is 1 missing value' if num_of_missing_values_in_dataset == 1
    else f'are {num_of_missing_values_in_dataset} missing values'} in the dataset.</h2>"""))

## 5. How many rows and columns are there in our Dataset?

In [24]:
ecommerce_purchases_df_shape = ecommerce_purchases_df.shape
print(f"Rows:    {ecommerce_purchases_df_shape[0]}\nColumns: {ecommerce_purchases_df_shape[1]}")

Rows:    10000
Columns: 14


## 6. Show all columns first then find the length of them.

In [25]:
TAB_SEPARATOR_LENGTH = 25

print(ecommerce_purchases_df.columns.to_list())
print("\nLength of columns", sep='')
print(f"\n{'column':<{TAB_SEPARATOR_LENGTH}}# of rows\n")
for column in ecommerce_purchases_df.columns:
    print(f"{column:.<{TAB_SEPARATOR_LENGTH}}{len(ecommerce_purchases_df[column])}")

['address', 'lot', 'time_am', 'browser_info', 'company', 'credit_card', 'cc_exp_date', 'cc_security_code', 'cc_provider', 'email', 'job', 'ip_address', 'language', 'purchase_price']

Length of columns

column                   # of rows

address..................10000
lot......................10000
time_am..................10000
browser_info.............10000
company..................10000
credit_card..............10000
cc_exp_date..............10000
cc_security_code.........10000
cc_provider..............10000
email....................10000
job......................10000
ip_address...............10000
language.................10000
purchase_price...........10000


## 7. Show all number of rows

In [26]:
print('Number of rows:', len(ecommerce_purchases_df))

Number of rows: 10000


## 8. Show complete information of dataset

In [27]:
def print_df_info(df):
    print('Type of dataset:', type(df))
    print('\nIs the dataset empty?:', df.empty)
    print('\nAxes:\n', df.axes, sep='')
    print('\nAttributes:\n', df.attrs, sep='')
    print('\nFlags:\n', df.flags, sep='')
    print('\nShape: (rows, columns)\n', df.shape, sep='')
    print('\nData Types:')
    display(df.dtypes)
    print('\nDescriptive Statistics:')
    display(df.describe().round(2))

In [28]:
print_df_info(ecommerce_purchases_df)

Type of dataset: <class 'pandas.core.frame.DataFrame'>

Is the dataset empty?: False

Axes:
[RangeIndex(start=0, stop=10000, step=1), Index(['address', 'lot', 'time_am', 'browser_info', 'company', 'credit_card',
       'cc_exp_date', 'cc_security_code', 'cc_provider', 'email', 'job',
       'ip_address', 'language', 'purchase_price'],
      dtype='object')]

Attributes:
{}

Flags:
<Flags(allows_duplicate_labels=True)>

Shape: (rows, columns)
(10000, 14)

Data Types:


address             string[python]
lot                 string[python]
time_am                    boolean
browser_info        string[python]
company             string[python]
credit_card                  Int64
cc_exp_date         string[python]
cc_security_code             Int64
cc_provider         string[python]
email               string[python]
job                 string[python]
ip_address          string[python]
language            string[python]
purchase_price             Float64
dtype: object


Descriptive Statistics:


credit_card  cc_security_code  purchase_price
count             10000.0           10000.0         10000.0
mean   2341373942652652.0            907.22           50.35
std    2256102885223829.0           1589.69           29.02
min         60401860543.0               0.0             0.0
25%      30563223611130.5             280.0           25.15
50%     869994171540613.5             548.0            50.5
75%    4492297929116673.5             816.0           75.77
max    6011999761615928.0            9993.0           99.99

In [29]:
ecommerce_purchases_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   address           10000 non-null  string 
 1   lot               10000 non-null  string 
 2   time_am           10000 non-null  boolean
 3   browser_info      10000 non-null  string 
 4   company           10000 non-null  string 
 5   credit_card       10000 non-null  Int64  
 6   cc_exp_date       10000 non-null  string 
 7   cc_security_code  10000 non-null  Int64  
 8   cc_provider       10000 non-null  string 
 9   email             10000 non-null  string 
 10  job               10000 non-null  string 
 11  ip_address        10000 non-null  string 
 12  language          10000 non-null  string 
 13  purchase_price    10000 non-null  Float64
dtypes: Float64(1), Int64(2), boolean(1), string(10)
memory usage: 1.0 MB


## 9. Show Highest and Lowest Purchase Prices

### a. Try to show all columns names so that one can easily pick the column

In [30]:
print(ecommerce_purchases_columns)

['address', 'lot', 'time_am', 'browser_info', 'company', 'credit_card', 'cc_exp_date', 'cc_security_code', 'cc_provider', 'email', 'job', 'ip_address', 'language', 'purchase_price']


In [31]:
PURCHASE_PRICE_COLUMN = 'purchase_price'

In [32]:
print(f"Number of rows with: min value: {len(ecommerce_purchases_df.loc[ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] == ecommerce_purchases_df[PURCHASE_PRICE_COLUMN].min()])},\
 max value: {len(ecommerce_purchases_df.loc[ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] == ecommerce_purchases_df[PURCHASE_PRICE_COLUMN].max()])}")


Number of rows with: min value: 2, max value: 2


In [33]:
ecommerce_purchases_df[PURCHASE_PRICE_COLUMN].loc[(ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] == ecommerce_purchases_df[PURCHASE_PRICE_COLUMN].max())
                           | (ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] == ecommerce_purchases_df[PURCHASE_PRICE_COLUMN].min())].sort_values()

2876      0.0
5487      0.0
2092    99.99
7807    99.99
Name: purchase_price, dtype: Float64

In [34]:
ecommerce_purchases_df.loc[(ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] == ecommerce_purchases_df[PURCHASE_PRICE_COLUMN].max())
                           | (ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] == ecommerce_purchases_df[PURCHASE_PRICE_COLUMN].min())].sort_values(PURCHASE_PRICE_COLUMN)

address    lot  time_am   
2876  332 Jones Parkways
East Katherineville, GA 64230  39 GT     True  \
5487     465 Mallory Ways
North Rebecca, RI 82734-1160  93 OH    False   
2092          63773 Shelton Greens
Ashleyton, MA 00493  56 lu     True   
7807              PSC 6177, Box 1004
APO AA 57143-1269  64 Nf     True   

                                           browser_info   
2876  Mozilla/5.0 (Macintosh; U; PPC Mac OS X 10_6_9...  \
5487  Mozilla/5.0 (X11; Linux i686; rv:1.9.6.20) Gec...   
2092  Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/53...   
7807  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_5_4)...   

                        company       credit_card cc_exp_date   
2876      Lyons, Diaz and Clark  4204500444841766       01/18  \
5487             Flynn and Sons    30469912089738       09/23   
2092                Pitts Group     4292741269160       06/18   
7807  Porter, Johnson and Pratt    30109394842259       11/25   

      cc_security_code    cc_provider                     email   
2876                38  VISA 16 digit      jennifer11@baker.com  \
5487               236       Discover       mjohnson@austin.org   
2092               824        Maestro  heatherwoodard@lloyd.com   
7807               918  VISA 16 digit         kelli72@gmail.com   

                         job       ip_address language  purchase_price  
2876    Biomedical scientist   223.47.177.174       it             0.0  
5487           Stage manager      43.99.56.59       zh             0.0  
2092  Surveyor, hydrographic  172.197.216.229       el           99.99  
7807     Surveyor, insurance     89.51.92.242       de           99.99

## 10. Show average purchase price

In [35]:
print(f'${round(ecommerce_purchases_df[PURCHASE_PRICE_COLUMN].mean(), 2)}')

$50.35


## 11. How many people have French 'fr' as their Language?

### a. First show all columns then all languages

In [36]:
print(ecommerce_purchases_columns)

['address', 'lot', 'time_am', 'browser_info', 'company', 'credit_card', 'cc_exp_date', 'cc_security_code', 'cc_provider', 'email', 'job', 'ip_address', 'language', 'purchase_price']


In [37]:
LANGUAGE_COLUMN = 'language'

### b. Check whether 'fr' is there or not

In [38]:
ecommerce_purchases_df[LANGUAGE_COLUMN].isin(['fr']).any()

True

### c. Compare language to 'fr'

In [39]:
print(f"{len(ecommerce_purchases_df.loc[ecommerce_purchases_df[LANGUAGE_COLUMN] == 'fr'])} people have French as their language.")

1097 people have French as their language.


## 12. Find Job Title Contains Engineer word

### a. Show all columns then look for 'Job' match it with 'engineer'

In [40]:
print(ecommerce_purchases_columns)

['address', 'lot', 'time_am', 'browser_info', 'company', 'credit_card', 'cc_exp_date', 'cc_security_code', 'cc_provider', 'email', 'job', 'ip_address', 'language', 'purchase_price']


In [41]:
JOB_COLUMN = 'job'

In [42]:
JOB_PROFESSION = 'engineer'

### b. Show all result rows having that word

In [43]:
ecommerce_purchases_df.loc[ecommerce_purchases_df[JOB_COLUMN].str.contains(JOB_PROFESSION)].head()

address    lot  time_am   
1    9374 Jasmine Spurs Suite 508
South John, TN 84...  28 rn    False  \
3                 7780 Julia Fords
New Stacy, WA 45798  36 vm    False   
50      41159 Michael Centers
Adamsfort, RI 37108-6674  46 Ce    False   
103                480 Ronald Cape
Johnville, MA 53644  11 wg     True   
193  1010 Mahoney Mills Suite 486
West Danielleview...  70 Fd     True   

                                          browser_info   
1    Opera/8.93.(Windows 98; Win 9x 4.90; en-US) Pr...  \
3    Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_0 ...   
50   Mozilla/5.0 (Windows 98; Win 9x 4.90; sl-SI; r...   
103  Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_5...   
193  Mozilla/5.0 (X11; Linux i686; rv:1.9.5.20) Gec...   

                             company       credit_card cc_exp_date   
1    Fletcher, Richards and Whitaker  3337758169645356       11/18  \
3    Williams, Marshall and Buchanan  6011578504430710       02/24   
50       Wright, Williams and Mendez  4008586485908075       05/19   
103                        Brown Inc    30276609779632       03/22   
193                    Wilson-Dennis    30359256148796       02/22   

     cc_security_code   cc_provider                          email   
1                 561    Mastercard             anthony41@reed.com  \
3                 384      Discover    brent16@olson-robinson.info   
50                945  JCB 16 digit      susanvalentine@obrien.org   
103               676      Discover  wtownsend@jackson-johnson.biz   
193                62  JCB 16 digit               stacy72@cruz.com   

                          job       ip_address language  purchase_price  
1           Drilling engineer     15.160.41.51       fr           70.73  
3           Drilling engineer     30.250.74.19       es           78.04  
50        Mechanical engineer  213.203.143.215       de           36.85  
103  Technical sales engineer   119.178.136.56       it           51.37  
193           Energy engineer   189.183.41.220       pt           96.79

### c. Now make your search as case insensitive

In [44]:
ecommerce_purchases_df.loc[ecommerce_purchases_df[JOB_COLUMN].str.lower().str.contains(JOB_PROFESSION)].tail()

address    lot  time_am   
9948  95544 Johnson Isle Suite 939
Michaelberg, RI 3...  91 bW     True  \
9952             9991 Vaughn Hills
Racheltown, PA 55409  36 KC    False   
9970               0060 Keith Stream
Westport, CO 47097  11 nt    False   
9977  02182 Keith Expressway
East Shannon, CT 20578-...  34 RL     True   
9996  832 Curtis Dam Suite 785
North Edwardburgh, TX...  41 JY     True   

                                           browser_info   
9948  Opera/8.36.(X11; Linux x86_64; sl-SI) Presto/2...  \
9952  Mozilla/5.0 (X11; Linux i686; rv:1.9.5.20) Gec...   
9970  Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_8...   
9977  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   
9996  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...   

                          company       credit_card cc_exp_date   
9948                 Fox-Peterson     4762924304307       03/17  \
9952     Ward, Smith and Castillo  6011679271321726       09/19   
9970     Carpenter, Good and Hart  6011485664704662       07/19   
9977  Deleon, Jacobson and Benton  4186094003664688       06/21   
9996     Hale, Collins and Wilson   210033169205009       07/25   

      cc_security_code   cc_provider                    email   
9948               567    Mastercard  haleybenjamin@gmail.com  \
9952               964       Voyager  jonesjennifer@olson.com   
9970               543      Discover  rangelbrian@hotmail.com   
9977               397  JCB 16 digit   daltoncarter@yahoo.com   
9996               207  JCB 16 digit       mary85@hotmail.com   

                      job      ip_address language  purchase_price  
9948  Structural engineer   120.36.140.58       en           71.89  
9952     Engineer, energy   116.228.12.42       es           39.63  
9970  Electrical engineer    242.8.85.205       en           17.76  
9977  Biomedical engineer   146.238.118.2       fr           94.93  
9996      Energy engineer  121.133.168.51       pt           25.63

### d. Find the length of resulting data too

In [45]:
TAB_SEPARATOR_LENGTH = max(len('Length using case-sensitive engineer:'), len('Length using case-insensitive engineer:'))

In [46]:
print(f"{'Length using case-sensitive engineer:':<{TAB_SEPARATOR_LENGTH}} {len(ecommerce_purchases_df.loc[ecommerce_purchases_df[JOB_COLUMN].str.contains(JOB_PROFESSION)])}\n\
{'Length using case-insensitive engineer:':<{TAB_SEPARATOR_LENGTH}} {len(ecommerce_purchases_df.loc[ecommerce_purchases_df[JOB_COLUMN].str.lower().str.contains(JOB_PROFESSION)])}")

Length using case-sensitive engineer:   531
Length using case-insensitive engineer: 984


## 13. Find The Email of the person with the following IP Address: 132.207.160.22

### a. First display all column to know on which column you have to work

In [47]:
ecommerce_purchases_columns

['address',
 'lot',
 'time_am',
 'browser_info',
 'company',
 'credit_card',
 'cc_exp_date',
 'cc_security_code',
 'cc_provider',
 'email',
 'job',
 'ip_address',
 'language',
 'purchase_price']

In [48]:
EMAIL_COLUMN = 'email'
IP_ADDRESS_COLUMN = 'ip_address'
IP_ADDRESS = '132.207.160.22'

### b. Compare the column to ip address and show the results

In [49]:
ecommerce_purchases_df.loc[ecommerce_purchases_df[IP_ADDRESS_COLUMN] == IP_ADDRESS]

address    lot  time_am   
2  Unit 0065 Box 5052
DPO AP 27450  94 vE    False  \

                                        browser_info   
2  Mozilla/5.0 (compatible; MSIE 9.0; Windows NT ...  \

                      company   credit_card cc_exp_date  cc_security_code   
2  Simpson, Williams and Pham  675957666125       08/19               699  \

    cc_provider                           email                       job   
2  JCB 16 digit  amymiller@morales-harrison.com  Customer service manager  \

       ip_address language  purchase_price  
2  132.207.160.22       de            0.95

### c. Now display email only of that IP address

In [50]:
ip_address_email_df = ecommerce_purchases_df[EMAIL_COLUMN].loc[ecommerce_purchases_df[IP_ADDRESS_COLUMN] == IP_ADDRESS]

display(ip_address_email_df.to_frame())
[email_address] = ip_address_email_df.to_list()
email_address

email
2  amymiller@morales-harrison.com

'amymiller@morales-harrison.com'

## 14. How many People have Mastercard as their Credit Card Provider and made a purchase above 50 dollars?

### a. First display columns

In [51]:
ecommerce_purchases_columns

['address',
 'lot',
 'time_am',
 'browser_info',
 'company',
 'credit_card',
 'cc_exp_date',
 'cc_security_code',
 'cc_provider',
 'email',
 'job',
 'ip_address',
 'language',
 'purchase_price']

In [52]:
CREDIT_CARD_PROVIDER_COLUMN = 'cc_provider'
CREDIT_CARD_PROVIDER = 'Mastercard'
PURCHASE_PRICE_LIMIT = 50

### b. Do some comparisons and show numbers

In [53]:
len(ecommerce_purchases_df.loc[(ecommerce_purchases_df[CREDIT_CARD_PROVIDER_COLUMN] == CREDIT_CARD_PROVIDER) &
                           (ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] > PURCHASE_PRICE_LIMIT)])

405

In [54]:
len(ecommerce_purchases_df.loc[(ecommerce_purchases_df[CREDIT_CARD_PROVIDER_COLUMN].str.lower() == CREDIT_CARD_PROVIDER.lower()) &
                           (ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] > PURCHASE_PRICE_LIMIT)])

405

In [55]:
len(ecommerce_purchases_df.loc[(ecommerce_purchases_df[CREDIT_CARD_PROVIDER_COLUMN].str.lower() == CREDIT_CARD_PROVIDER.lower()) &
                           (ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] >= PURCHASE_PRICE_LIMIT)])

405

Comparing Mastercard purchases with less than $50 purchases, in which there are more as compared to greater than or equal to $50 (411 vs 405)

In [56]:
len(ecommerce_purchases_df.loc[(ecommerce_purchases_df[CREDIT_CARD_PROVIDER_COLUMN].str.lower() == CREDIT_CARD_PROVIDER.lower()) &
                           (ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] < PURCHASE_PRICE_LIMIT)])

411

In [57]:
ecommerce_purchases_df.loc[(ecommerce_purchases_df[CREDIT_CARD_PROVIDER_COLUMN].str.lower() == CREDIT_CARD_PROVIDER.lower()) &
                           (ecommerce_purchases_df[PURCHASE_PRICE_COLUMN] > PURCHASE_PRICE_LIMIT)].head()

address    lot  time_am   
1   9374 Jasmine Spurs Suite 508
South John, TN 84...  28 rn    False  \
18  461 Christopher Square
West Michaelchester, CA...  17 SB    False   
31                    USNS Alvarado
FPO AA 27052-1231  26 Lh    False   
35  93392 Webb Gardens Apt. 220
Laurabury, AR 9993...  37 om     True   
90         431 Bowen Lights
Fergusonborough, MH 01362  31 tG     True   

                                         browser_info   
1   Opera/8.93.(Windows 98; Win 9x 4.90; en-US) Pr...  \
18  Mozilla/5.0 (X11; Linux i686; rv:1.9.6.20) Gec...   
31  Opera/8.84.(X11; Linux i686; sl-SI) Presto/2.9...   
35  Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_7...   
90  Mozilla/5.0 (Windows NT 5.2; it-IT; rv:1.9.1.2...   

                            company       credit_card cc_exp_date   
1   Fletcher, Richards and Whitaker  3337758169645356       11/18  \
18           Beard, Abbott and Pena  6011350184276270       12/22   
31                  Nicholson Group     4614997834548       03/22   
35                         Mora Ltd  6011049630969815       09/16   
90                     Copeland-Lee      639032576097       06/18   

    cc_security_code cc_provider                   email   
1                561  Mastercard      anthony41@reed.com  \
18               767  Mastercard      hannah63@yahoo.com   
31               909  Mastercard    ashley12@hotmail.com   
35               367  Mastercard  hgonzalez@mcdowell.com   
90               868  Mastercard       nrogers@brown.com   

                                       job      ip_address language   
1                        Drilling engineer    15.160.41.51       fr  \
18                            Photographer  73.250.176.201       el   
31                         Sales executive  94.176.142.201       en   
35             Lecturer, further education    216.23.95.40       zh   
90  Teacher, English as a foreign language  153.188.13.203       es   

    purchase_price  
1            70.73  
18           70.15  
31           94.14  
35           97.46  
90           78.86

## 15. Find the email of the person with the following Credit Card Number: 4664825258997302

In [58]:
ecommerce_purchases_columns

['address',
 'lot',
 'time_am',
 'browser_info',
 'company',
 'credit_card',
 'cc_exp_date',
 'cc_security_code',
 'cc_provider',
 'email',
 'job',
 'ip_address',
 'language',
 'purchase_price']

In [59]:
CREDIT_CARD_NUMBER_COLUMN = 'credit_card'
CREDIT_CARD_NUMBER = 4664825258997302

In [60]:
credit_card_number_email = ecommerce_purchases_df[EMAIL_COLUMN].loc[ecommerce_purchases_df[CREDIT_CARD_NUMBER_COLUMN] == CREDIT_CARD_NUMBER]

display(credit_card_number_email.to_frame())
[email_address] = credit_card_number_email.to_list()
email_address

email
9992  bberry@wright.net

'bberry@wright.net'

## 16. How many people purchase during the AM and how many people purchase during PM?

In [61]:
ecommerce_purchases_columns

['address',
 'lot',
 'time_am',
 'browser_info',
 'company',
 'credit_card',
 'cc_exp_date',
 'cc_security_code',
 'cc_provider',
 'email',
 'job',
 'ip_address',
 'language',
 'purchase_price']

In [62]:
AM_PM_COLUMN = 'time_am'

In [63]:
print(f"{len(ecommerce_purchases_df.loc[ecommerce_purchases_df[AM_PM_COLUMN]])} people purchase during the AM and \
{len(ecommerce_purchases_df.loc[~ecommerce_purchases_df[AM_PM_COLUMN]])} people purchase during the PM.") 

4932 people purchase during the AM and 5068 people purchase during the PM.


## 17. How many people have a credit card that expires in 2020?

### a. Show columns first

In [64]:
ecommerce_purchases_columns

['address',
 'lot',
 'time_am',
 'browser_info',
 'company',
 'credit_card',
 'cc_exp_date',
 'cc_security_code',
 'cc_provider',
 'email',
 'job',
 'ip_address',
 'language',
 'purchase_price']

In [65]:
CREDIT_CARD_DATE_COLUMN = 'cc_exp_date'
YEAR_OF_EXPIRY = '20'

### b. Check the format of date

In [66]:
ecommerce_purchases_df[CREDIT_CARD_DATE_COLUMN]

0       02/20
1       11/18
2       08/19
3       02/24
4       10/25
        ...  
9995    03/22
9996    07/25
9997    05/21
9998    11/17
9999    02/19
Name: cc_exp_date, Length: 10000, dtype: string

### c. Create a user defined function which counts how many persons having credit card expired in 2020

In [67]:
def count_credit_card_expired(df, year):
    return len(df.loc[df.str[-2:] == year])

### d. Iterate the date using for loop and split date and count it

In [68]:
count_expired_cards = 0
for row in ecommerce_purchases_df[CREDIT_CARD_DATE_COLUMN]:
    date_result = row.split('/')
    if date_result[1] == YEAR_OF_EXPIRY:
        count_expired_cards += 1

### e. Show total sum of count

In [69]:
print(f"Total sum of expired count: {count_expired_cards}")

Total sum of expired count: 988


### f. You can use any other method too

In [70]:
print(f"Total sum of expired count: {count_credit_card_expired(ecommerce_purchases_df[CREDIT_CARD_DATE_COLUMN], YEAR_OF_EXPIRY)}")

Total sum of expired count: 988


## 18. What are the top 5 most popular email providers (e.g. Gmail.com, yahoo.com, etc...)

### a. Create either a user defined function or use your own any other way for getting the result

In [71]:
def top_values_from_df(df, numbers_to_rank, column_to_compare):
    return df.groupby([column_to_compare]).size().sort_values(ascending=False).head(numbers_to_rank)

In [72]:
def top_values_from_series(series, numbers_to_rank):
    return series.value_counts().head(numbers_to_rank)

In [73]:
ecommerce_purchases_email_provider = ecommerce_purchases_df.copy()
ecommerce_purchases_email_provider[EMAIL_COLUMN] = ecommerce_purchases_df[EMAIL_COLUMN].str.split('@').str[1]

In [74]:
top_values_from_df(ecommerce_purchases_email_provider, 5, EMAIL_COLUMN).to_frame(name='count')

count
email              
hotmail.com    1638
yahoo.com      1616
gmail.com      1605
smith.com        42
williams.com     37

In [75]:
top_values_from_series(ecommerce_purchases_email_provider[EMAIL_COLUMN], 5).to_frame()

count
email              
hotmail.com    1638
yahoo.com      1616
gmail.com      1605
smith.com        42
williams.com     37

## Bonus: Export dataframe to desired location with no index column

### Data Visualization/Analysis Additions Before Exporting

In [76]:
ecommerce_purchases_df['email_provider'] = ecommerce_purchases_df[EMAIL_COLUMN].str.split('@').str[1]

In [77]:
language_dict = {
    'de': "German", 'el': "Greek", "en": "English", "es": "Spanish",
    "fr": "French", "it": "Italian", "pt": "Portuguese", 
    "ru": "Russian", "zh": "Chinese"
}

In [78]:
ecommerce_purchases_df[LANGUAGE_COLUMN] = ecommerce_purchases_df[LANGUAGE_COLUMN].replace(language_dict)

In [79]:
ecommerce_purchases_df['state'] = ecommerce_purchases_df['address'].str.split().str[-2]

In [80]:
states_list = [ 'AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA',
           'HI', 'IA', 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA', 'MD', 'ME',
           'MI', 'MN', 'MO', 'MS', 'MT', 'NC', 'ND', 'NE', 'NH', 'NJ', 'NM',
           'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX',
           'UT', 'VA', 'VT', 'WA', 'WI', 'WV', 'WY']

In [81]:
states_dictionary = { 
    'AK': 'Alaska', 'AL': 'Alabama', 'AR': 'Arkansas',
    'AZ': 'Arizona', 'CA': 'California', 'CO': 'Colorado', 'CT': 'Connecticut',
    'DC': 'District of Columbia', 'DE': 'Delaware', 'FL': 'Florida', 'GA': 'Georgia',
    'HI': 'Hawaii', 'IA': 'Iowa', 'ID': 'Idaho', 'IL': 'Illinois', 'IN': 'Indiana',
    'KS': 'Kansas', 'KY': 'Kentucky', 'LA': 'Louisiana', 'MA': 'Massachusetts',
    'MD': 'Maryland', 'ME': 'Maine', 'MI': 'Michigan', 'MN': 'Minnesota',
    'MO': 'Missouri', 'MS': 'Mississippi', 'MT': 'Montana', 'NC': 'North Carolina',
    'ND': 'North Dakota', 'NE': 'Nebraska', 'NH': 'New Hampshire', 'NJ': 'New Jersey',
    'NM': 'New Mexico', 'NV': 'Nevada', 'NY': 'New York', 'OH': 'Ohio',
    'OK': 'Oklahoma', 'OR': 'Oregon', 'PA': 'Pennsylvania', 'RI': 'Rhode Island',
    'SC': 'South Carolina', 'SD': 'South Dakota', 'TN': 'Tennessee', 'TX': 'Texas',
    'UT': 'Utah', 'VA': 'Virginia', 'VT': 'Vermont', 'WA': 'Washington',
    'WI': 'Wisconsin', 'WV': 'West Virginia', 'WY': 'Wyoming'
}

In [82]:
american_territory_replacement = {'AS': 'American Samoa', 'VI': "Virgin Islands", 'GU': "Guam", "PR": "Puerto Rico",
    'MH': "Marshall Islands", 'PW': 'Palau', 'AE': "Armed Forces Europe", "AA": "Armed Forces America", "AP": "Armed Forces Pacific",
    "FM": "Micronesia", "MP": "Northern Mariana Islands"
}

In [83]:
american_territory_latitude_replacement = {
    'MH': "7.116667", 'PW': '7.5', 'AE': "52.52", "AA": "38.895", "AP": "37.56",
    "FM": "6.916", "MP": "15.19"
}

In [84]:
american_territory_longitude_replacement = {
    'MH': "171.066667", 'PW': '134.616667', 'AE': "13.405", "AA": "-77.036", "AP": "126.99",
    "FM": "158.183", "MP": "145.74"
}

In [85]:
ecommerce_purchases_df['territory'] = ecommerce_purchases_df['state'].where(~ecommerce_purchases_df['state'].isin(states_list), None)

In [86]:
ecommerce_purchases_df['state'] = ecommerce_purchases_df['state'].where(ecommerce_purchases_df['state'].isin(states_list), None)

In [87]:
ecommerce_purchases_df['state'] = ecommerce_purchases_df['state'].replace(states_dictionary)
ecommerce_purchases_df['territory_lat'] = ecommerce_purchases_df['territory'].replace(american_territory_latitude_replacement)
ecommerce_purchases_df['territory_long'] = ecommerce_purchases_df['territory'].replace(american_territory_longitude_replacement)
ecommerce_purchases_df['territory'] = ecommerce_purchases_df['territory'].replace(american_territory_replacement)

In [88]:
ecommerce_purchases_df.columns = ecommerce_purchases_df.columns.str.replace('_', ' ').str.title()

### Export Cleaned Dataframe to File

In [89]:
ecommerce_purchases_df.to_csv('ecommerce_purchases_clean.csv', index=False)